Source/WTF/wtf/Vector.h

```diff
@@ -714,8 +714,17 @@ class Vector : private VectorBuffer<T, inlineCapacity, Malloc> {
 
         asanSetInitialBufferSizeTo(size);
 
-        for (size_t i = 0; i < size; ++i)
-            unsafeAppendWithoutCapacityCheck(valueGenerator(i));
+        if constexpr (std::is_same_v<std::invoke_result_t<Functor, size_t>, std::optional<T>>) {
```

Comment:
Should there be a `shrinkToFit` somewhere? Seems unclear if we should keep extra capacity indefinitely. I’d expect this to be the same as `compactMap` in that respect.

Reply:
This is something that bugs me with compactMap(). Shouldn't we leave this up to the caller? If the caller wants to store the vector somewhere for an extended period of time, they can call shrinkToFit().

In this instance, the only call site doesn't really want to call shrinkToFit() as the vector is on the stack and doesn't live long (the previous code didn't shrinkToFit()). If I force a shrinkToFit() in the constructor then this call site will pay a cost unnecessarily.

Reply:
Given that was would be your preference? My plan was to not call it and leave it up to the caller (The current caller doesn't call it because it would only hurt performance without benefit AFAIK). Alternatively, we could add a parameter to decide whether or not to call shrinkToFit() and make it the default behavior (optional parameter).

Reply:
Sure that makes sense.

Getting this resizing optimal is tricky and I wish it wasn't. In older code the most common mistake was keeping a vector around for a long time with excess capacity and @smfr tried to change our idioms to decrease the chance of that mistake.

But clearly **this** call site doesn't want a `shrinkToFit`. I'm just not sure how we can make it easy to get right. If we are keeping a vector around for a while we want it right-sized and I can easily imagine someone using this constructor for that kind of situation in the future. For now I guess it's OK to be inconsistent with `compactMap` while we think about the best future idiom.